controllers/clustergroupupgrade_controller.go

```diff
@@ -194,282 +213,332 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
+	}
 
-		readyCondition := meta.FindStatusCondition(clusterGroupUpgrade.Status.Conditions, "Ready")
+	r.updateStatus(ctx, clusterGroupUpgrade)
 
-		if readyCondition == nil {
-			meta.SetStatusCondition(&clusterGroupUpgrade.Status.Conditions, metav1.Condition{
-				Type:    "Ready",
-				Status:  metav1.ConditionFalse,
-				Reason:  "UpgradeNotStarted",
-				Message: "The ClusterGroupUpgrade CR is not enabled",
-			})
-			nextReconcile = requeueImmediately()
-		} else if readyCondition.Status == metav1.ConditionFalse {
-			if readyCondition.Reason == "PrecachingRequired" {
-				nextReconcile = requeueWithLongInterval()
-			} else if readyCondition.Reason == "UpgradeNotStarted" || readyCondition.Reason == utils.CannotStart {
-				// Before starting the upgrade check that all the managed policies exist.
-				var allManagedPoliciesExist bool
-				var managedPoliciesInfo policiesInfo
-				allManagedPoliciesExist, managedPoliciesInfo, err =
-					r.doManagedPoliciesExist(ctx, clusterGroupUpgrade, true)
-				if err != nil {
+	backupCondition := meta.FindStatusCondition(clusterGroupUpgrade.Status.Conditions, string(utils.ConditionTypes.BackupSuceeded))
+	precachingCondition := meta.FindStatusCondition(clusterGroupUpgrade.Status.Conditions, string(utils.ConditionTypes.PrecachingSuceeded))
+	suceededCondition := meta.FindStatusCondition(clusterGroupUpgrade.Status.Conditions, string(utils.ConditionTypes.Succeeded))
+
+	if suceededCondition != nil {
+		if suceededCondition.Status == metav1.ConditionTrue {
+			// Upgrade has successfully finished
+			if clusterGroupUpgrade.Status.Status.CompletedAt.IsZero() {
+				r.Log.Info("Upgrade is completed")
+				// Take actions after upgrade is completed
+				clusterGroupUpgrade.Status.Status.CurrentBatch = 0
+				clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt = metav1.Time{}
+				if err = r.takeActionsAfterCompletion(ctx, clusterGroupUpgrade); err != nil {
 					return
 				}
+				// Set completion time only after post actions are executed with no errors
+				clusterGroupUpgrade.Status.Status.CompletedAt = metav1.Now()
+			}
+		} else {
+			// Upgrade has failed
+			r.Recorder.Event(clusterGroupUpgrade, corev1.EventTypeWarning, "UpgradeTimedOut", "The ClusterGroupUpgrade CR policies are taking too long to complete")
+			r.Log.Info("CGU has timed out")
+			r.addClustersStatusOnTimeout(clusterGroupUpgrade)
+			// On timeout we don't want to complete actions other then to delete the resources
+			err = r.deleteResources(ctx, clusterGroupUpgrade)
+			if err != nil {
+				return
+			}
 
-				if allManagedPoliciesExist {
-					// Build the upgrade batches.
-					err = r.buildRemediationPlan(ctx, clusterGroupUpgrade, managedPoliciesInfo.presentPolicies)
-					if err != nil {
-						return
-					}
+		}
+	} else if (backupCondition == nil || (backupCondition != nil && (backupCondition.Reason == string(utils.ConditionReasons.Completed) || backupCondition.Reason == string(utils.ConditionReasons.PartiallyDone)))) &&
```

Comment:
Is it better to add the condition with a reason like "not enabled" so that this expression can be a bit simpler?

Reply:
Or at least wrap this up in a bool function with an easy-to-read name.